src/python/pants/backend/project_info/depgraph.py

```diff
@@ -0,0 +1,200 @@
+# Copyright 2021 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from __future__ import annotations
+
+import json
+import logging
+from dataclasses import dataclass
+from enum import Enum
+from json import JSONEncoder
+from typing import Iterable, cast
+
+from pants.backend.python.target_types import PythonRequirementsField
+from pants.base.specs import AddressSpecs, MaybeEmptyDescendantAddresses
+from pants.engine.console import Console
+from pants.engine.goal import Goal, GoalSubsystem, LineOriented
+from pants.engine.internals.selectors import MultiGet
+from pants.engine.rules import Get, collect_rules, goal_rule, rule
+from pants.engine.target import Dependencies, DependenciesRequest, Targets
+from pants.util.meta import frozen_after_init
+
+logger = logging.getLogger(__name__)
+
+
+class Granularity(Enum):
+    FILE = "file"
+    TARGET = "target"
+    DIRECTORY = "directory"
+
+
+class GraphFormat(Enum):
+    JSON = "json"
+    DOT = "dot"
+
+
+class DepGraphSubsystem(LineOriented, GoalSubsystem):
+    name = "depgraph"
+    help = "Emit a dependency graph for the entire repo."
+
+    @classmethod
+    def register_options(cls, register):
+        super().register_options(register)
+        register(
+            "--granularity",
+            type=Granularity,
+            default=Granularity.FILE,
+            help="Emit the graph at this level of granularity",
+        )
+        register(
+            "--format",
+            type=GraphFormat,
+            default=GraphFormat.JSON,
+            help="Emit the graph in this format.",
+        )
+
+    @property
+    def format(self) -> GraphFormat:
+        return cast(GraphFormat, self.options.format)
+
+    @property
+    def granularity(self) -> Granularity:
+        return cast(Granularity, self.options.granularity)
+
+
+class DepGraphGoal(Goal):
+    subsystem_cls = DepGraphSubsystem
+
+
+@dataclass(frozen=True)
+class Vertex:
+    id: str
+    data: dict
+
+    def to_dict(self) -> dict:
+        return {"id": self.id, "data": self.data}
+
+
+@dataclass(frozen=True)
+class Edge:
+    src_id: str
+    dep_id: str
+    data: dict
+
+    def to_dict(self) -> dict:
+        return {"src_id": self.src_id, "dep_id": self.dep_id, "data": self.data}
```

Comment:
I suggest using [dataclasses.asdict](https://docs.python.org/3/library/dataclasses.html#dataclasses.asdict) as it will produce the some output as this code.

alternatively, get rid of those methods (on both classes) and just have the json encoder call asdict...